src/misc/cycle_bases.c

```diff
@@ -0,0 +1,507 @@
+                    }
+
+                    upe = VECTOR(pred_edge)[up];
+                    IGRAPH_CHECK(igraph_vector_int_push_back(&u_back, upe));
+                    up = IGRAPH_OTHER(graph, upe, up);
+
+                    if (up == vp) {
+                        break;
+                    }
+
+                    vpe = VECTOR(pred_edge)[vp];
+                    IGRAPH_CHECK(igraph_vector_int_push_back(&v_back, vpe));                    
+                    vp = IGRAPH_OTHER(graph, vpe, vp);
+                }
+
+                u_back_len = igraph_vector_int_size(&u_back);
+                v_back_len = igraph_vector_int_size(&v_back);
+                IGRAPH_CHECK(igraph_vector_int_init(&cycle, u_back_len + v_back_len));
+                IGRAPH_FINALLY(igraph_vector_int_destroy, &cycle);
+
+                for (j=0; j < v_back_len; ++j) {
+                    VECTOR(cycle)[j] = VECTOR(v_back)[j];
+                }
+                for (j=0; j < u_back_len; ++j) {
+                    VECTOR(cycle)[v_back_len + j] = VECTOR(u_back)[u_back_len - j - 1];
+                }
+
+                igraph_vector_int_clear(&v_back);
+                igraph_vector_int_clear(&u_back);
+
+                IGRAPH_CHECK(igraph_vector_int_list_push_back(result, &cycle));
+                IGRAPH_FINALLY_CLEAN(1); /* pass ownership of 'cycle' to 'result' */
+            } else {
+                /* encountering u for the first time, queue it for processing */
+
+                /* Only queue vertices with distance at most 'cutoff' from the root. */
+                /* Negative 'cutoff' indicates no cutoff. */
+                if (cutoff < 0 || vdist < cutoff) {
+                    igraph_dqueue_int_push(&q, u);
+                    igraph_dqueue_int_push(&q, vdist + 1);
+                    VECTOR(*visited)[u] = mark + 1;
+                    VECTOR(pred_edge)[u] = e;
+                }
+            }            
+        }
+
+        VECTOR(*visited)[v] = mark + 2; /* mark v as processed */
+    } /* ! igraph_dqueue_int_empty(&q) */
+
+    igraph_dqueue_int_destroy(&q);
+    igraph_vector_int_destroy(&v_back);
+    igraph_vector_int_destroy(&u_back);
+    igraph_vector_int_destroy(&pred_edge);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/**
+ * \function igraph_fundamental_cycles
+ * \brief Finds a fundamental cycle basis.
+ *
+ * This function computes a fundamental cycle basis associated with a breadth-first
+ * search tree of the graph.
+ *
+ * </para><para>
+ * Edge directions are ignored. Multi-edges and self-loops are supported.
+ *
+ * \param graph The graph object.
+ * \param start_vid If negative, a complete fundamental cycle basis is returned.
+ *   If a vertex ID, the fundamental cycles associated with the BFS tree rooted
+ *   in that vertex will be returned, only for the weakly connected component
+ *   containing that vertex.
+ * \param cutoff If negative, a complete cycle basis is returned. Otherwise, only
```

Comment:
Could we define this parameter in a way that it represents the maximum length of the cycles we are looking for? This would be more intuitive from the end user's point of view.

Reply:
Especially in higher-level interfaces where the autocompletion of the IDE might show you that you have a "cutoff" parameter and then you simply assume "oh, so it's the maximum length of the cycles we are looking for".

Reply:
I did think about this of course, and it would look more natural to use a cutoff on cycle length. However, I'm not sure that's a good idea right now. That is part of the reason why I want to make the API `\experimental`.

This is feature is completely novel, as far as I am aware. No other graph library implements it, nor have I seen it described. It gives a massive speedup, effectively descreasing the exponent by one in the complexity of the algorithms, when the graph satsifies certain conditions. I think it's better to be explicit about how it works, i.e. that it is the BFS depth. Also, we can only limit the cycle lengths to odd numbers.

Reply:
Then let's call it `bfs_depth_limit` for the time being and then it's explicit that it's not the cycle length that is being limited here.

Reply:
Renamed to `bfs_cutoff` (keep it short).